crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -562,33 +564,104 @@ impl SnapshotRepository {
     }
 }
 
-async fn create_object_store_client(destination: Url) -> anyhow::Result<Arc<dyn ObjectStore>> {
+async fn create_object_store_client(
+    destination: Url,
+    options: &SnapshotsOptions,
+) -> anyhow::Result<Arc<dyn ObjectStore>> {
     // We use the AWS SDK configuration and credentials provider so that the conventional AWS
     // environment variables and config files work as expected. The object_store crate has its
     // own configuration mechanism which doesn't support many of the AWS conventions. This
     // differs quite a lot from the Lambda invoker which uses the AWS SDK, and that would be a
     // very surprising inconsistency for customers. This mechanism allows us to infer the region
     // and securely obtain session credentials without any hard-coded configuration.
     let object_store: Arc<dyn ObjectStore> = if destination.scheme() == "s3" {
-        let builder = if let Ok(profile) = std::env::var("AWS_PROFILE") {
-            // Infer the AWS region from the profile, or use AWS_REGION if set.
-            let region = aws_config::load_defaults(BehaviorVersion::v2024_03_28())
-                .await
-                .region()
-                .context("Unable to determine AWS region")?
-                .to_string();
-            debug!(?profile, ?region, "Using AWS SDK credentials provider");
-            AmazonS3Builder::new()
-                .with_credentials(Arc::new(AwsSdkCredentialsProvider {
-                    credentials_provider: DefaultCredentialsChain::builder().build().await,
-                }))
-                .with_region(region)
+        let builder = match &options.aws_profile {
+            Some(profile) => {
+                debug!(profile, "Using AWS profile for snapshot repository access");
+                let sdk_config = aws_config::defaults(BehaviorVersion::v2024_03_28())
+                    .profile_name(profile)
+                    .load()
+                    .await;
+
+                let region = DefaultRegionChain::builder()
+                    .profile_name(profile)
+                    .build()
+                    .region()
+                    .await
+                    .context("Unable to determine region from profile")?;
```

Comment:
What if I have configured a region via `aws_region`? Or is it invalid to configure a profile w/o a region?

Reply:
This was incorrect, thank you for pointing this out! Restate config region should be used, if set.

You can definitely have a profile config without a region; AWS will generally default to `us-east-1` for historic reasons so I thought this is just to satisfy the compiler, but actually we can get `None` back. I had a global default region in my local config which is why I never hit that path testing :-)

This should be better now!